swarm/src/connection.rs

```diff
@@ -358,7 +361,16 @@ where
                         *shutdown = Shutdown::Later(Delay::new(dur), t)
                     }
                 }
-                (_, KeepAlive::No) => *shutdown = Shutdown::Asap,
+                (_, KeepAlive::No) if idle_timeout == &Duration::ZERO => {
+                    *shutdown = Shutdown::Asap;
+                }
+                (Shutdown::Later(_, _), KeepAlive::No) => {
+                    // Do nothing, i.e. let the shutdown timer continue to tick.
+                }
+                (_, KeepAlive::No) => {
+                    let deadline = Instant::now() + *idle_timeout;
+                    *shutdown = Shutdown::Later(Delay::new(*idle_timeout), deadline);
+                }
```

Comment:
If I understand the match arm further above correctly (copied below for ease of read) the following situation can happen. The user sets `idle_connection_timeout` to a large value, say 1h. One `ConnectionHandler` returns `KeepAlive::Until` with a short timeout, say 1s. In case nothing else happens on the connection, it will be closed after 1s.

One could argue that this is the correct behavior. I don't think this is intuitive.

``` rust
                (Shutdown::Later(timer, deadline), KeepAlive::Until(t)) => {
                    if *deadline != t {
                        *deadline = t;
                        if let Some(dur) = deadline.checked_duration_since(Instant::now()) {
                            timer.reset(dur)
                        }
                    }
                }
```

Should we deprecate `KeepAlive::Until` with this pull request already, to prevent the above footgun to exit for long?

Reply:
I think deprecating `Until` takes a bit more work, at least if we want to remove all usages in our codebase at the same time.

Should we have the 1h start ticking once the `Until` finishes?

Reply:
> Should we have the 1h start ticking once the `Until` finishes?

Instead of keeping the connection alive for the duration of the sum of the two (i.e. `KeepAlive::Until` + `idle_connection_timeout`) , how about for the duration of the maximum of the two.

From both perspectives (i.e. `KeepAlive::Until` and `idle_connection_timeout`), the connection is kept alive at least as much as configured, potentially longer.

I would find this more intuitive and I assume this is easier to implement, as it would pretty much only require changing a `!=` for a `<`.

@thomaseizinger thoughts?

Reply:
Yeah that sounds good to me! :)

I'll fix that later so hopefully we can land this soon 😍